RTClib.cpp

```diff
@@ -746,6 +746,168 @@ TimeSpan::TimeSpan(int16_t days, int8_t hours, int8_t minutes, int8_t seconds)
 /**************************************************************************/
 TimeSpan::TimeSpan(const TimeSpan &copy) : _seconds(copy._seconds) {}
 
+/*!
+    @brief  Create a new TimeSpan from an iso8601 formatted period string
+
+    If the string is entirely malformed (doesn't begin with P or -) this will
+    construct a TimeSpan of 0 seconds. If parsing fails before the end of the
+    string, this constuctor will interpret its argument as the longest
+    well-formed prefix string.
+
+    For example, the string P5M10~ will parse the same as P5M
+
+    @param iso8601 the formatted string
+*/
+TimeSpan::TimeSpan(char *iso8601) : _seconds(0) {
+  int32_t sign = 1;
+  char *cursor = iso8601;
+  if (*cursor == '-') {
+    sign = -1;
+    cursor++;
+  }
+  if (*cursor == 'P') {
+    cursor++;
+  } else {
+    return;
+  }
+
+  char *rest;
+  long num = strtol(cursor, &rest, 10);
+  cursor = rest;
+  if (*cursor == 'D') {
+    _seconds += sign * SECONDS_PER_DAY * num;
+    cursor++;
+  }
+
+  if (*cursor == 'T') {
+    cursor++;
+    num = strtol(cursor, &rest, 10);
+    cursor = rest;
+  } else if (*cursor == '\0') {
+    return;
+  } else {
+    // error: malformed iso8601
+    return;
+  }
+
+  if (*cursor == 'H') {
+    _seconds += sign * SECS_PER_MIN * MINS_PER_HOUR * num;
+    cursor++;
+    num = strtol(cursor, &rest, 10);
+    cursor = rest;
+  }
+
+  if (*cursor == 'M') {
+    _seconds += sign * SECS_PER_MIN * num;
+    cursor++;
+    num = strtol(cursor, &rest, 10);
+    cursor = rest;
+  }
+
+  if (*cursor == 'S') {
+    _seconds += sign * num;
+  }
+}
+
+/*!
+    @brief  Formats this TimeSpan according to iso8601
+
+    See toCharArray for more details on the format
+
+    @return String of formatted TimeSpan
+*/
+String TimeSpan::toString() const {
+  constexpr size_t buflen = 19;
+  char buf[buflen];
+  this->toCharArray(buf, buflen);
+  return String(buf);
+}
+
+/*!
+    @brief  Formats this TimeSpan according to iso8601
+
+    Fails (returning 0) if buf is too small to store the result. buf size 19
+    will never fail.
+    Example: TimeSpan t(32, 23, 54, 11) formats to "P32DT23H54M11S".
+
+    @param buf char array to write output. Always contains trailing '\0'.
+    @param len the length of buf. Must be at least 1.
+    @return number of bytes written excluding trailing '\0' or 0 on failure
+*/
+size_t TimeSpan::toCharArray(char *buf, size_t len) const {
+  size_t written = 0;
+
+  if (_seconds == 0) {
+    written = snprintf(buf, len, "PT0S");
+    if (written >= len) {
+      buf[len - 1] = '\0';
```

Comment:
The multiple instances of this statement can be factored into a single one at the beginning of the function. Then `snprintf()` should write no more than `len-1-written` bytes in order to not overwrite that `'\0'`.

Reply:
But then this library would eagerly overwrite the last byte of the user's array even when they only use the first 5 (of say 19) bytes. This doesn't seem highly problematic, but does violates the least surprise principle. On the other hand, the code bloat is probably significant from all those `if`s. I changed it to use only one if at the end.